[Backend Receipts] Add backend vs local receipt eligibility usecase checks

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes #11827
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
This PR introduces an eligibility check for backend receipts, so the option will be available for sites running WooCommerce 8.7.0+ (not launched yet).

We render either the legacy or the new `See receipt` based  the WooCommerce version, and if there are existing receipts already saved on device (only for paid orders): 
* If there are existing receipts on device, we just follow the existing logic and display locally-generated receipts.
* If there are no existing receipts on device, we display the new `See receipt` for eligible stores, in order to fetch one from the backend.

## Changes
- Updated `[Debug] See backend receipt` to `See Receipt`, and made them mutually exclusive: The row will either show a link to the old receipt, to the new one, or no row.
- Added an eligibility checker, we only show the new logic if: locally-generated receipt doesn't yet exist, feature flag is enabled, and WC version is supported. 

## Testing instructions
* On any store with a current WooCommerce version, confirm that the legacy flow works as expected, that is: An unpaid order shows no "See receipt" row in order details. When paying for the order via a card present method, saving the receipt, and refreshing the order (either navigate back and in again, or use pull to refresh) shows the "See receipt" row. Tapping the row, displays the old receipt (screenshot for comparison below)
* On an eligible store (you can use https://mywootestingstore.mystagingwebsite.com/): 
  * Go to an unpaid order (eg: "Pending Payment"), and observe that we do not render a `See receipt` row.
  * Go to an paid order (eg: "Completed"), and observe that `See receipt` row is rendered and a backend receipt is shown (screenshot for comparison below).

## Screenshots
| Has local receipt | Eligible for backend receipt | Awaiting payment |
|--------|--------|--------|
| ![Simulator Screenshot - iPhone 15 - 2024-02-02 at 10 13 03](https://github.com/woocommerce/woocommerce-ios/assets/3812076/857fa3b6-3c65-4389-bb25-65176b69dd98) | ![simulator_screenshot_B629D3E8-A20A-4EF9-B3A1-E4943DD0D83B](https://github.com/woocommerce/woocommerce-ios/assets/3812076/e6bb04c7-c548-4ef6-b648-6a387b623f12) | ![2024-02-02 awaiting payment](https://github.com/woocommerce/woocommerce-ios/assets/3812076/37840bf6-6174-487c-81c0-4f078a227190) |

Note that there's no summary description under `Paid` for new receipts, this is expected since currently the `paymentSummary` performs a check on payment methods to see if a card present was used. We'll attempt to improve this separately. Logged at: https://github.com/woocommerce/woocommerce-ios/issues/11878.

| Old receipt | New receipt |
|--------|--------|
| ![Simulator Screenshot - iPhone 15 - 2024-02-02 at 12 02 00](https://github.com/woocommerce/woocommerce-ios/assets/3812076/529b7259-029d-4595-b8f4-7ac17265e4b5) | ![Simulator Screenshot - iPhone 15 - 2024-02-02 at 12 01 11](https://github.com/woocommerce/woocommerce-ios/assets/3812076/c27dc5cb-41fd-4278-a775-33c09e2f5c19) |

iamgabrielma committed:
Add ReceiptEligibilityUseCase

iamgabrielma committed:
Merge branch 'trunk' into issue/11827-backend-receipts-eligibility

iamgabrielma committed:
Update woo minimum version

iamgabrielma committed:
Unit tests for ReceiptEligibilityUseCase

iamgabrielma committed:
typo

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/183">17.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Delete outdated test

We no longer show the receipt row for all cases (needs to be paid orders) so having the feature flag enabled is not enough to render it

iamgabrielma committed:
Check for paid order before rendering receipt row

iamgabrielma committed:
Update receipt ui text

iamgabrielma commented on WooCommerce/Classes/ViewModels/Order Details/Receipts/ReceiptEligibilityUseCase.swift:
```diff
@@ -0,0 +1,49 @@
+    enum Constants {
+        static let wcPluginName = "WooCommerce"
+        static let wcPluginMinimumVersion = "8.7.0"
+        static let wcPluginDevVersion = "8.6.0-dev-7625495467-gf50cc6b"
```
The `8.6.0-dev-7625495467-gf50cc6b` version is the latest that worked correctly in our test environment, and the one used on the  https://mywootestingstore.mystagingwebsite.com/ test site. If you're using a different test site with a different WooCommerce dev version please update this string.

This will be removed once the feature goes live, initially planned with WC 8.7

iamgabrielma commented:
👋 Only one review is needed. There's no rush on reviewing this today, next week is fine.

iamgabrielma commented on WooCommerce/Classes/ViewModels/Order Details/Receipts/ReceiptEligibilityUseCase.swift:
```diff
@@ -0,0 +1,49 @@
+        
+        let action = SystemStatusAction.fetchSystemPlugin(siteID: siteID, systemPluginName: Constants.wcPluginName) { wcPlugin in
+            // 1. WooCommerce must be installed and active
+            guard let wcPlugin = wcPlugin, wcPlugin.active else {
```
I don't think we really need the `.active` check here, I mostly added it for return early consistency like we do with other plugin checks, since Woo could be inactive but the user being previously logged in the app.

bozidarsevo commented:
Shouldn't "1.00$" be "$1.00"? I know that US dollars and some other currencies show the symbol before the amount/number?

bozidarsevo commented:
I noticed now, did not catch that before, maybe we could change a bit logic when we tap on the "See Receipt". Now we call:
`ReceiptAction.retrieveReceipt`
and if the internet connection is slow nothing happens, we just wait and then when we get it we push the view controller. Maybe we should push it immediately and load it there, and show a loading indicator?

iamgabrielma commented:
> Shouldn't "1.00$" be "$1.00"? I know that US dollars and some other currencies show the symbol before the amount/number?

Agree. I'm not sure if the latest version of the dev plugin already has tackled this, since we've sent some feedback regarding the receipt, but I'll check with the backend folks to see if they're aware of this 👍

iamgabrielma commented:
> I noticed now, did not catch that before, maybe we could change a bit logic when we tap on the "See Receipt". Now we call: `ReceiptAction.retrieveReceipt` and if the internet connection is slow nothing happens, we just wait and then when we get it we push the view controller. Maybe we should push it immediately and load it there, and show a loading indicator?

Yeah! There's a logged issue to handle the waiting better: https://github.com/woocommerce/woocommerce-ios/issues/11857 , I'll do this along error handling here: https://github.com/woocommerce/woocommerce-ios/issues/11829 since we plan to show something to the merchant while they wait or if the loading fails

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/183">17.2</a>. The due date for this milestone has already passed.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
